packages/opentelemetry-resources/src/platform/node/detectors/AwsEcsDetector.ts

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Resource } from '../../../Resource';
+import { CONTAINER_RESOURCE } from '../../../constants';
+import { Detector } from '../../../types';
+import { ResourceDetectionConfigWithLogger } from '../../../config';
+import * as util from 'util';
+import * as fs from 'fs';
+import * as os from 'os';
+
+export const cache = {
+  readFileAsync: util.promisify(fs.readFile),
+};
+
+/**
+ * The AwsEcsDetector can be used to detect if a process is running in AWS
+ * ECS and return a {@link Resource} populated with data about the ECS
+ * plugins of AWS X-Ray. Returns an empty Resource if detection fails.
+ */
+class AwsEcsDetector implements Detector {
+  readonly CONTAINER_ID_LENGTH = 64;
+  readonly DEFAULT_CGROUP_PATH = '/proc/self/cgroup';
+  readonly ECS_METADATA_KEY_V4 = 'ECS_CONTAINER_METADATA_URI_V4';
+  readonly ECS_METADATA_KEY_V3 = 'ECS_CONTAINER_METADATA_URI';
+
+  async detect(config: ResourceDetectionConfigWithLogger): Promise<Resource> {
+    if (
+      !process.env.ECS_CONTAINER_METADATA_URI_V4 &&
+      !process.env.ECS_CONTAINER_METADATA_URI
+    ) {
+      config.logger.debug('AwsEcsDetector failed: Process is not on ECS');
+      return Resource.empty();
+    }
+
+    let hostName, containerId;
+    try {
+      hostName = os.hostname();
+    } catch (e) {
+      config.logger.warn(
+        `AwsEcsDetector failed to read host name: ${e.message}`
+      );
+    }
+    try {
+      const rawData = await cache.readFileAsync(
+        this.DEFAULT_CGROUP_PATH,
+        'utf8'
+      );
+      containerId = this._getContainerId(rawData);
+    } catch (e) {
+      config.logger.warn(
+        `AwsEcsDetector failed to read container ID: ${e.message}`
+      );
+    }
+
+    return !hostName && !containerId
+      ? Resource.empty()
+      : new Resource({
+          [CONTAINER_RESOURCE.NAME]: hostName || '',
+          [CONTAINER_RESOURCE.ID]: containerId || '',
+        });
+  }
+  /**
+   * Read container ID from cgroup file
+   * In ECS, even if we fail to find target file
+   * or target file does not contain container ID
+   * we do not throw an error but throw warning message
+   * and then return null string
+   */
+  private _getContainerId(rawData: string): string {
```

Comment:
I still think a better way to detect the current docker container ID would be preferrable. Not a blocker if you disagree.

```typescript
import * as path from "path";
import * as fs from "fs";
import * as util from "util";

const readFileAsync = util.promisify(fs.readFile);

async function getContainerid() {
  const cpuset = await readFileAsync('/proc/1/cpuset');
  return path.basename(cpuset);
}
```

Reply:
Yes, I have found a related discussion here https://stackoverflow.com/questions/60605500/how-to-access-container-id-in-nodejs-application-in-host-network-mode
Since I am also not an expert of docker and ECS, I am currently ask other AWS engineers for suggestions.
But I currently think '/proc/self/cgroup' is more ECS-specific.

Reply:
After I tried to build a EC2 instance and ran a container to see the exact condition these days. I now believe either way works very well. But to be consistent with Java implementation, I prefer to maintain the current 'proc/self/cgroup' implementation. @dyladan